thorlcr/graph/thgraphmaster.cpp

```diff
@@ -1609,45 +1612,31 @@ void CJobMaster::broadcast(ICommunicator &comm, CMessageBuffer &msg, mptag_t mpt
         replyTag = queryJobChannel(0).queryMPServer().createReplyTag();
         msg.setReplyTag(replyTag);
     }
-    if (globals->getPropBool("@broadcastSendAsync", true)) // only here in case of problems/debugging.
+    auto f = [&](unsigned r)
     {
-        class CSendAsyncfor : public CAsyncFor
+        if (!aborting || comm.verifyConnection(r, (unsigned)-1, false))
         {
-            CMessageBuffer &msg;
-            mptag_t mptag;
-            unsigned timeout;
-            StringAttr errorMsg;
-            ICommunicator &comm;
-        public:
-            CSendAsyncfor(ICommunicator &_comm, CMessageBuffer &_msg, mptag_t _mptag, unsigned _timeout, const char *_errorMsg)
-                : comm(_comm), msg(_msg), mptag(_mptag), timeout(_timeout), errorMsg(_errorMsg)
-            {
-            }
-            void Do(unsigned i)
-            {
-                if (!comm.send(msg, i+1, mptag, timeout))
-                    throw createBCastException(i+1, errorMsg);
-            }
-        } afor(comm, msg, mptag, timeout, errorMsg);
-        try
-        {
-            afor.For(groupSizeExcludingMaster, groupSizeExcludingMaster);
-        }
-        catch (IException *e)
-        {
-            EXCLOG(e, "broadcastSendAsync");
-            abort(e);
-            throw;
+            if (!comm.send(msg, r, mptag, timeout))
+                throw createBCastException(r, errorMsg);
         }
+    };
+    std::vector<std::future<void>> sendResultsFuture;
+    for (unsigned dstrank=1; dstrank<=groupSizeExcludingMaster; ++dstrank)
+        sendResultsFuture.push_back(std::async(f, dstrank));
+    try
+    {
+        for (auto & sendResultFuture: sendResultsFuture)
+            sendResultFuture.get();
```

Comment:
FYI, if get isn't called for one of the std::future object, then it may cause the child thread to continue to exist until the parent thread is terminated.  Do you think this may cause a problem?

Reply:
good catch, yes.
I guess add a try/catch around the get() call, and keep the 1st IException and rethrow it after the loop